### PR TITLE
Aerocore2

### PR DESCRIPTION
**Pull request by jmcwil**
Converting Aerocore2 to CMake system.  Current version will compile, but still gets stuck at waiting for the bootloader.

@LorenzMeier Thanks for taking a look at this for me!

Relates to issue #4827

### CODE REVIEW

File: nuttx-configs/aerocore/nsh/defconfig [hunk 1]
```diff
@@ -371,7 +348,6 @@ CONFIG_BOOT_RUNFROMFLASH=y
 #
 # Board Selection
 #
-CONFIG_ARCH_BOARD_AEROCORE=y
```
Comment (davids5) [inline nuttx-configs/aerocore/nsh/defconfig]: This line needs to be restored 

and add 
CONFIG_ARCH_BOARD="aerocore"

Then if you run make menuconfig or make oldconfig -, once you have copied these back to the defconfig restore these lines:

CONFIG_ARCH_BOARD="aerocore"
CONFIG_ARCH_BOARD_AEROCORE=y

File: nuttx-configs/aerocore/nsh/defconfig [hunk 2]
```diff
@@ -371,7 +348,6 @@ CONFIG_BOOT_RUNFROMFLASH=y
...
 CONFIG_ARCH_BOARD_CUSTOM=y
```
Comment (davids5) [inline nuttx-configs/aerocore/nsh/defconfig]: Delete this line

File: aerocore_build_errors_fixed [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// COPIED FILE OVER FROM PX4/FIRMWARE
```
Comment (dagar) [inline aerocore_build_errors_fixed]: remove this file

File: cmake/configs/aerocore_build_7.log [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Script started on Mon 27 Jun 2016 11:38:10 AM EDT
```
Comment (dagar) [inline cmake/configs/aerocore_build_7.log]: remove this file

File: mavlink/include/mavlink/v2.0 [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit 1e2920c68df46c25b7a14ca6cc47a478c6081d81
+Subproject commit 699a2b0df1b259a962a01af788e8306ee8d016d8
```
Comment (dagar) [inline mavlink/include/mavlink/v2.0]: Revert the submodule change.

File: src/drivers/ms5611/ms5611.cpp [hunk 6]
```diff
@@ -829,6 +829,8 @@ struct ms5611_bus_option {
 	{ MS5611_BUS_SPI_EXTERNAL, "/dev/ms5611_spi_ext", &MS5611_spi_interface, PX4_SPI_BUS_EXT, NULL },
 #endif
 #ifdef PX4_SPIDEV_BARO
+//try changing to fix compiling error
+//#if defined(PX4_SPIDEV_BARO) && defined(PX4_SPI_BUS_BARO)
```
Comment (dagar) [inline src/drivers/ms5611/ms5611.cpp]: remove

File: src/include/mavlink/mavlink_log.h [hunk 7]
```diff
@@ -0,0 +1,164 @@
+/****************************************************************************
```
Comment (dagar) [inline src/include/mavlink/mavlink_log.h]: remove

File: src/modules/systemlib/param/param.c [hunk 8]
```diff
@@ -65,6 +65,7 @@
 
 #include "uORB/uORB.h"
 #include "uORB/topics/parameter_update.h"
+//Cannot find this file anywhere!!
```
Comment (dagar) [inline src/modules/systemlib/param/param.c]: remove

**Comment by dagar** [comment]
Looking good. Fix the formatting (install astyle and run make check_format) and the builds will start passing. Add aerocore to the qgc_firmware target in the Makefile and it will start compiling on travis-ci.
https://github.com/PX4/Firmware/blob/master/Makefile#L310

File: src/drivers/boards/aerocore/aerocore_init.c [hunk 9]
```diff
@@ -209,18 +209,20 @@ static struct spi_dev_s *spi4;
 
 #include <math.h>
```
Comment (davids5) [inline src/drivers/boards/aerocore/aerocore_init.c]: Delete from 210-225

File: src/drivers/boards/aerocore/aerocore_timer_config.c [hunk 10]
```diff
@@ -42,7 +42,8 @@
 #include <stdint.h>
 
 #include <stm32.h>
-#include <stm32_gpio_out.h>
+//#include <stm32_gpio_out.h> //almost positive this is wrong. Added below statement
```
Comment (davids5) [inline src/drivers/boards/aerocore/aerocore_timer_config.c]: delete

**Comment by LorenzMeier** [comment]
@jmcwil Have you seen the feedback from @dagar and @davids5. You will want to make sure to get this in quickly so that the board config is maintained by the upstream team.

**Comment by jmcwil** [comment]
@dagar @davids5 Thank you for all of the feedback!!  I believe that I have fixed the errors that you specified.  The only one that may still be giving me trouble could be the mavlink/v2.0 problem.  Can you help me figure out what is going on with it?  I am not an experienced git user yet, and I think I may have messed something up with mavlink.

While the code continues to compile on my host machine, I have still be unsuccessful in getting it to upload to my device.  Does this make sense given the errors you were able to detect?

Thanks again for all of your help!  I really appreciate it.

**Comment by dagar** [comment]
To fix the mavlink version cd into the folder and checkout the same version of the submodule that's in master.

    cd mavlink/include/mavlink/v2.0/
    git checkout 1e2920c

Then go back to the top level of Firmware and commit that change.

I'm not sure what's going on with upload. Maybe the id is wrong somewhere? You can force it to get around that error, but I'm not sure if that's a good idea until we understand the actual problem.

    python Tools/px_uploader.py --force --port /dev/ttyACM0 build_aerocore-v2_default/src/firmware/nuttx/nuttx-aerocore-v2-default.px4

I guessed the path to the firmware, so it's probably wrong. Just locate for the .px4 file.

**Comment by LorenzMeier** [comment]
@jmcwil Did you set up the permissions in Linux as advised on http://dev.px4.io?

If the board can be really be detected it will tell you that you're trying to upload to the wrong board:
```
Found board b,0 bootloader rev 5 on /dev/tty.usbmodem1
WARNING: Firmware not suitable for this board (board_type=11 board_id=9)
```

**Comment by LorenzMeier** [comment]
If you don't get this error message then your hardware is not actually showing up on USB or permissions are wrong. Can you please share the output of ```lsusb``` and uninstall ```modemmanager```?

**Comment by jmcwil** [comment]
@dagar Thank you!  That helped a lot.

@LorenzMeier I will double check that the permissions are correct.  I have already uninstalled modem manager.  
The output of lsusb detects the Aerocore: ````jessica@mcwiljl-ll1:~$ lsusb
Bus 004 Device 002: ID 8087:0024 Intel Corp. Integrated Rate Matching Hub
Bus 004 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 003 Device 002: ID 8087:0024 Intel Corp. Integrated Rate Matching Hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 004: ID 413c:3012 Dell Computer Corp. Optical Wheel Mouse
Bus 001 Device 003: ID 413c:2010 Dell Computer Corp. Keyboard
Bus 001 Device 002: ID 413c:1003 Dell Computer Corp. Keyboard Hub
Bus 001 Device 009: ID 26ac:0011  
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
````  It appears that the Aerocore is showing up as bus 001 Device 009.

So far, the only way I can successfully flash my device us through QGC.  This allows me to access the nsh through a serial connection.  However, only the default version(v1.3.4) allows me to access the nsh from my aerocore after flashing, and I need to be able to modify the firmware. However, when I try to upload anything from terminal, I am not getting the wrong device error that you posted.

Thanks for your time!

**Comment by LorenzMeier** [comment]
@jmcwil The aerocore does report as a Pixhawk, which is wrong:
Bus 001 Device 009: ID 26ac:0011

The uploader script should still pick it up. I don't see why it would not and QGC does.

**Comment by jmcwil** [comment]
@LorenzMeier Interesting, thanks for explaining that.  Could that be from flashing in QGC?  What should be the correct output for aerocore?

**Comment by LorenzMeier** [comment]
@jmcwil Can you check what you get for ```ls /dev/serial/by-id/*``` ? I suspect the uploader script is not finding your board and you need to do a change equivalent to this:
https://github.com/PX4/Firmware/commit/f8382a27137b15da1ede6689ba5213abe48c809e

**Comment by jmcwil** [comment]
@LorenzMeier Thanks for the reply.  Here is the current output.  This is after flashing PX4 v1.3.4 from QGC.
`/dev/serial/by-id/usb-Gumstix_AeroCore_0-if00
`
A confusing thing to note is that I have been able to flash non-functional versions of the firmware, versions that will successfully upload, but will not allow me to access the NuttX Shell.  If I run ls /dev/serial/by-id after flashing one of these versions, I will get the output ` ls: cannot access /dev/serial/by-id/: No such file or directory`.  Any idea what could cause this particular firmware-dependent behavior?

**Comment by LorenzMeier** [comment]
This means that the USB interface is not coming up. You need to connect a serial (like real serial, not USB-serial) console to work out why the boot fails.

**Comment by jmcwil** [comment]
@LorenzMeier Thanks for pointing this out.  Would you mind explaining what hardware I need to accomplish this?  I have never heard of doing this before and I was unable to find a resource that explains how to do this with Gumstix.

**Comment by davids5** [comment]
@jmcwil 

Do you have a link for the HW you are using - I can review the schematic and tell you what you will need and how to connect it.

Assuming the IO is 3.3 Volt you will need one of these http://www.digikey.com/product-detail/en/ftdi-future-technology-devices-international-ltd/TTL-232R-3V3/768-1015-ND/1836393

**Comment by davids5** [comment]
@jmcwil 
This may be easier to hook up but does not have full Hardware HS (TRS/CTS) - it iwll work for debug just not other things.

http://media.digikey.com/pdf/Data%20Sheets/Olimex%20LTD%20PDFs/USB-SERIAL-CABLE-F_Web.pdf

**Comment by jmcwil** [comment]
@davids5 Is this going to be similar to the set up shown [here](http://dev.px4.io/advanced-system-console.html) for setting up a serial connection to pixhawk without a dronecode probe?  If so, then I guess I just need to locate the Tx, Rx, and Gnd pins on the [Aerocore2](https://store.gumstix.com/aerocore-2-for-duovero.html).  

I guess I am still a little confused how this will be different from using the serial port on the Aerocore.  From what I read in the documentation for Aerocore, it seems to me that the conversion to FTDI is implemented on the board, so a serial connection can be achieved at the microusb. 

Thanks for your continued support!

**Comment by davids5** [comment]
Are you on Skype? 

I am david_s5 - We can discuss this in real time

**Comment by jmcwil** [comment]
Unfortunately I am not available for a Skype call at this time.

**Comment by davids5** [comment]
Well the short answer is:

Debugging a fresh board bring up is complicated.

The serial console IO is simple (not alot has to work) to get it going. Especially on the version of Nuttx from master.

> conversion  to FTDI is implemented on the board, so a serial connection can be achieved at the microusb.

That is not how it works - (I can explain it on a call).

since CONFIG_UART7_SERIAL_CONSOLE=y we need to hook up to 
image

On a std  TTL-232R-3V3 cable Orange is TX from PC to RX on board, and Yellow is RX of the PC connected to TX of the board.

2.4 Cortex-M4 UART 7
Pin 1 is TX - Yellow (RX)  TTL-232R-3V3
Pin 2 is GND - Black (GND) TTL-232R-3V3
Pin 3 is RX  -  Orange  (TX) TTL-232R-3V3

**Comment by davids5** [comment]
@jmcwil - Any progress?

**Comment by jmcwil** [comment]
@davids5 Thanks for checking in with me!  I have not gotten a chance to set this up yet, but I plan to take a look at it at the beginning of next week.

Thanks for explaining what connections are necessary.  Also, thank you for offering to Skype with me.  I would really appreciate the opportunity to speak in real time sometime next week. Your patience with my inexperience has been very helpful and greatly appreciated.

**Comment by jmcwil** [comment]
@LorenzMeier @davids5  Thanks for your patience and explanation of how to get a real serial connection to the board. I wired an FTDI cable to UART7, and I am now able to connect to the NuttX Shell on UART7 with `sudo screen /dev/ttyUSB0 115200 8N1`.  Could you please tell me how I can use this connection to debug this upload?  I tried running the upload with the NuttX Shell open, but that did not give me any outputs.  I am unsure how to proceed.

**Comment by jmcwil** [comment]
@LorenzMeier @davids5 I did some searching and found out I should try ver all.  Here is the output.
`nsh> ver all
HW arch: AEROCORE
FW git-hash: ba6b821efc417db91751b6459fc250774b170f6f
Build datetime: Jun 17 2016 20:17:33
Toolchain: 4.7.4 20130913 (release) [ARM/embedded-4_7-branch revision 202601]
MCU: STM32F42x, rev. Y

WARNING   WARNING   WARNING!
Revision Y has a silicon errata
This device can only utilize a maximum of 1MB flash safely!
http://px4.io/help/errata

UID: 2A002B:34324714:30333837
`
From what I have read, this is a hardware issue that you guys have seen before.  Is there a way to fix this for Aerocore?

Thanks for all of your help!

**Comment by davids5** [comment]
@jmcwil 

Great Progress!

The fact NSH is running and there is no crash dump on the screen is a good sign.

The flash size warning is not an issue. 

What was programmed on the unit? ba6b821efc417db91751b6459fc250774b170f6f does not appear in your commit list, and the date is a month old?

Is this supposed to be your build programed with QGC?

Please type ```reboot``` and send along the output.

Are there LED's on this board?

**Comment by jmcwil** [comment]
@davids5 The output was obtained after flashing from QGC.

After typing reboot, I get this output (possibly more before that but I do not know that I can catch that part).

```
nsh: blinkm: command not found
ver hwcmp match: AEROCORE
 0:   SYS_AUTOSTART: match
[i] No autostart
nsh: test: syntax error
nsh: test: syntax error
dataman: Unknown restart, data manager file '/fs/mtd_dataman' size is 101504 bytes
[MS5611_SPI] on SPI bus 3 at 2 (11000 KHz)
<adc> init done
ver hwcmp match: AEROCORE
[L3GD20] on SPI bus 3 at 1 (11000 KHz)
[LSM303D] on SPI bus 3 at 0 (11000 KHz)
nsh: meas_airspeed: command not found
nsh: ets_airspeed: command not found
nsh: ets_airspeed: command not found
nsh: px4flow: command not found
nsh: ll40ls: command not found
commander: No RGB LED found at /dev/rgbled0
telem> PREFLIGHT FAIL: MAG #0 UNCALIBRATED
telem> PREFLIGHT FAIL: ACCEL #0 UNCALIBRATED
telem> PREFLIGHT FAIL: GYRO #0 UNCALIBRATED
telem> PREFLIGHT FAIL: AIRSPEED SENSOR MISSING
[fmuservo] default PWM output device
<fmuservo> MODE_8PWM
<fmuservo> set_pwm_rate 0 50 50
<fmuservo> adjusted actuator update interval to 20ms
[i] FMU mode_pwm started
ver hwcmp match: AEROCORE
mavlink_if0: mode: 0, data rate: 5000 B/s on /dev/ttyS1 @ 57600B
mavlink_if0: hardware flow control not supported
ver hwcmp match: AEROCORE
mavlink_if1: mode: 0, data rate: 1000 B/s on /dev/ttyS2 @ 115200B
mavlink_if1: hardware flow control not supported
param: Error: Parameter UAVCAN_ENABLE not found.
[i] No autostart ID found
[i] No microSD card found
navigator: offboard mission updated: dataman_id=0, count=0, current_seq=0

NuttShell (NSH)
nsh> commander: INVAL: ARMING_STATE_STANDBY_ERROR - ARMING_STATE_STANDBY
```
I tried flashing my version of the firmware for aerocore through QGC and I now I am unable to connect to the NuttX Shell when it is flashed.  When I open the serial connection it is just blank.  Is there a way to find out why this happens?  Again, the output does not show that the aerocore is connected at all.  `ls /dev/serial/by-id` only gives usb-FTDI_TTL232R_FTH03ISL-if00-port0, the cable on UART7.


There are 3 LED's (blue, red, orange).  When a valid version (default from QGC) is loaded, the LED's are all off.  However, when my version is loaded, the Blue LED is on solid.  Perusing the Aerocore forums has given indication that this often means the bootloader is missing or invalid; however, my bootloader has been working to flash the QGC firmware.  The board also has a buzzer, and it makes a solid beep while this version is loaded.  When a valid version is loaded, it has a pattern of beeps that it makes.  It seems to me that the bootloader is having trouble loading the firmware properly, but I am not sure how to verify or correct this.

**Comment by davids5** [comment]
Is there a reset button on the HW? If so push it when the console is connected and see if there is any output.

Compare the PX4 ld.script and Bootloader source LD script and insure that have the same load locations.

The next thing is either Nuttx auto_led or Jtag debugging. Auto LED will require you to read up on Nuttx and add a file to turn on/off the LED. Jtag will require JTAG HW

**Comment by jmcwil** [comment]
@davids5 There is no output when I hit the reset button.  The LED's flash to indicate that the board is powering up, but then the Blue LED goes on solid again.

I can check the ld.scripts.

**Comment by davids5** [comment]
Can you get on Skype today?

**Comment by jmcwil** [comment]
@davids5 Thanks for offering to Skype, I don't take this opportunity for granted.  I am having a little trouble getting my development machine to connect right now.  Do I need to have my hardware on hand for the call, or are we just discussing the situation in more detail?  If not, it will be better for me to use my other machine later today, but I would not have my device with me.  If I need my hardware on hand I can try again with this machine.  Please let me know what I need and what your availability is.

**Comment by davids5** [comment]
@jmcwil 

No need for HW

I will be in all day, so any time is fine.

**Comment by Pietrohl** [comment]
Helo, 

I am working with aerocore 2, and I am also following the development Guide for the PX4. I was able to flash the firmware on the board using different versions of QgroundControl, but I could not make any of the firmware versions work properly (2.7.1 was able to connect to the board but freezes soon after,  2.9.0 and 2.9.6 was not able to connect to the board alone, but could calibrate some parameters and arm), and for what I could understand this is probably for a lack of compatibility due the fact that the development for the aerocore was discontinued (latest change on the aerocore github account is about 1 year old: https://github.com/aerocore).

I have used the latest firmware version for the aerocore 2 from the gumstix repositiory (https://github.com/aerocore/Firmware), but also have experienced problems to start Mavlink automaticaly, and to arm and takeof the Quadcopter. Also when using this firmware from gumstix QgroundControl gives an error saying that a few parameters are missing (see attached files), and because of that it does not work properly. I also have not been able to use Mission Planner since it shows a compatibility error with this Firmware version.

I'm glad that a firmware version for this board is been worked on, I will try to follow all the  modifications done until now and hope to be able to help with anything else. 

Since I have had so many problems using QGroundControl what would be the best version to use with the Aerocore 2.


screenshot from 2016-07-22 14 17 12
screenshot from 2016-07-22 14 18 12
screenshot from 2016-07-22 14 19 24

**Comment by dagar** [comment]
Did you select an airframe? The other errors are the result of QGC and Firmware version mismatch. With PX4 master you should use a QGC daily build.

**Comment by Pietrohl** [comment]
Yes I selected the airframe and did all the calibrations that I was able. I have used the PX4 master and the QGC does not recognize the Aerocore 2.

**Comment by LorenzMeier** [comment]
@davids5 Will you look into this?

**Comment by davids5** [comment]
@LorenzMeier 

The HW is aerocore2 we had aerocore, this PR is calling aerocore2 hw aerocore- it it ok to overwrite the aerocore(v1) with this or shall we preserve the original aerocore and split them out to v1 and v2?

**Comment by LorenzMeier** [comment]
Nobody will maintain v1, so you can as well rename it to v2 and strip it.

**Comment by davids5** [comment]
FYI: @jmcwil, @LorenzMeier 

Superceeded by https://github.com/PX4/Firmware/pull/5281

**Pull request closed without merging**